### PR TITLE
Add rollback functionality to asgard module.

### PR DESCRIPTION
**Pull request by doctoryes**
Add click script to rollback to ASGs. Add tests for rollback and failed rollback.

@edx/pipeline-team Please review.

### CODE REVIEW

File: tubular/asgard.py [hunk 1]
```diff
@@ -468,6 +466,32 @@ def elbs_for_asg(asg):
...
+    Returns:
+        dict(str, dict): Returns a dictionary with the keys: 'current_asgs' and 'disabled_asgs'
+
+    Raises:
+        TimeoutException: When the task to bring up the new instance times out.
+        BackendError: When the task to bring up the new instance fails.
+        ASGDoesNotExistException: If the ASG being queried does not exist.
+    """
+    try:
+        _red_black_deploy(rollback_to_clustered_asgs, current_clustered_asgs)
```
Comment (macdiesel) [inline tubular/asgard.py]: Before you attempt to rollback you should attempt to remove the delete_on_ts tag from the rollback_to_clustered_asgs. If the janitor runs while the process is in flight it may lead to some strange results. 

I would suggest removing the tag, checking to see if deletion is in progress then proceeding with the _red_black_deploy call.

File: tubular/asgard.py [hunk 2]
```diff
@@ -468,6 +466,32 @@ def elbs_for_asg(asg):
...
+        return {'current_asgs': rollback_to_clustered_asgs, 'disabled_asgs': current_clustered_asgs }
```
Comment (macdiesel) [inline tubular/asgard.py]: Question, prior to this change the current_asgs and disabled_asgs were generated by the code as the deployments actually happened. With this change we will now assume that this is actually the case. Do we want to make that assumption? 

It might be better to let _red_black_deploy actually tell us what it did.

File: tubular/asgard.py [hunk 3]
```diff
@@ -477,7 +501,7 @@ def deploy(ami_id):
         ami_id(str): AWS AMI ID
 
     Returns:
-        dict(str, [str]): Returns a dictionary with the keys: 'current_asgs' and 'disabled_asgs'
+        dict(str, dict): Returns a dictionary with the keys: 'current_asgs' and 'disabled_asgs'
```
Comment (macdiesel) [inline tubular/asgard.py]: Could we add the ami_id to the return dict structure please? This Information is necessary for the rollback call.

File: tubular/asgard.py (hunk 1 above)
Comment (doctoryes) [reply tubular/asgard.py]: Will do.

File: tubular/asgard.py (hunk 2 above)
Comment (doctoryes) [reply tubular/asgard.py]: This code was written with the assumption that `_red_black_deploy()` will either succeed completely _or_ will raise an exception. But, after reviewing the code, I see that's not necessarily the case. That means I'll need to move from an exception-raising pattern to a return of (success, asg_clusters_state) in order to return the state properly before raising an error in the caller.

tl;dr - will do.

File: tubular/asgard.py (hunk 3 above)
Comment (doctoryes) [reply tubular/asgard.py]: Will do.

**Comment by doctoryes** [comment]
@macdiesel Comments are addressed - ready for another review.

File: scripts/rollback_asg.py [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+            deploy_info = asgard.rollback(current_asgs, disabled_asgs, ami_id)
+        else:
+            click.echo('Would have triggered a rollback of {}'.format(ami_id))
+            deploy_info = {}
+
+        if out_file:
+            with open(out_file, 'w') as stream:
+                yaml.safe_dump(deploy_info, stream, default_flow_style=False, explicit_start=True)
+        else:
+            print yaml.dump(deploy_info, default_flow_style=False, explicit_start=True)
```
Comment (macdiesel) [inline scripts/rollback_asg.py]: safe_dump?

File: tubular/asgard.py [hunk 5]
```diff
@@ -490,78 +545,150 @@ def deploy(ami_id):
...
-    disabled_asg = defaultdict(list)
-    for cluster,asgs in existing_clusters.iteritems():
+    except Exception as wait_fail:
+        LOG.info("Some ASGs are failing ELB health checks. Disabling traffic to all new ASGs.")
+        for cluster, asgs in newly_enabled_asgs.iteritems():
+            for asg in asgs:
+                try:
+                    disable_cluster_asg(cluster, asg)
+                except:
+                    LOG.warning("Unable to disable ASG '{}' after waiting for healthy ELBs.".format(asg))
```
Comment (macdiesel) [inline tubular/asgard.py]: Should this continue here?

File: tubular/asgard.py [hunk 6]
```diff
@@ -490,78 +545,150 @@ def deploy(ami_id):
...
-    for cluster, asg in new_asgs.iteritems():
-        if is_asg_pending_delete(asg) or not is_asg_enabled(asg):
-            raise BackendError("New Autoscale Group {} is pending delete, Aborting the disabling of old ASGs.".format(asg))
-
-    disabled_asg = defaultdict(list)
-    for cluster,asgs in existing_clusters.iteritems():
+    except Exception as wait_fail:
+        LOG.info("Some ASGs are failing ELB health checks. Disabling traffic to all new ASGs.")
+        for cluster, asgs in newly_enabled_asgs.iteritems():
+            for asg in asgs:
```
Comment (macdiesel) [inline tubular/asgard.py]: NIT: We essentially run this logic twice for failures. It might be good to pull it out in to a local method like you did with enable_cluster_asg and disable_cluster_asg.

File: tubular/asgard.py [hunk 7]
```diff
@@ -490,78 +545,150 @@ def deploy(ami_id):
...
+                elbs_to_monitor.extend(elbs_for_asg(asg))
+                newly_enabled_asgs[cluster].append(asg)
+            except Exception:
+                LOG.error("Error enabling ASG '{}'. Disabling traffic to all new ASGs.".format(asg))
+                LOG.error(traceback.format_exc())
+                # Disable the ASG which failed first.
+                disable_cluster_asg(cluster, asg)
+                # Then disable any new other ASGs that have been newly enabled.
+                for cluster, asgs_to_disable in newly_enabled_asgs.iteritems():
+                    for asg in asgs_to_disable:
```
Comment (macdiesel) [inline tubular/asgard.py]: The other place we rollback the ASGs see nit about moving this to method.

File: tubular/asgard.py [hunk 8]
```diff
@@ -490,78 +545,150 @@ def deploy(ami_id):
...
+        LOG.info("Some ASGs are failing ELB health checks. Disabling traffic to all new ASGs.")
+        for cluster, asgs in newly_enabled_asgs.iteritems():
+            for asg in asgs:
+                try:
+                    disable_cluster_asg(cluster, asg)
+                except:
+                    LOG.warning("Unable to disable ASG '{}' after waiting for healthy ELBs.".format(asg))
+        return (False, asgs_enabled, asgs_disabled)
+
+    # Ensure the new ASGs are still healthy and not pending delete before disabling the old ASGs.
```
Comment (macdiesel) [inline tubular/asgard.py:682]: Something to consider in the future, perhaps we should have a timer here where we sleep for a (few) minute(s) before disabling the old ASGs? This way if something happens just after they get load that is causing them to fall out of the load balancer the deploy will fail.

@edx/devops any thoughts?

File: tubular/ec2.py [hunk 9]
```diff
@@ -168,6 +168,24 @@ def tag_asg_for_deletion(asg_name, seconds_until_delete_delta=1800):
...
+
+    Returns:
+        None
+    """
+    tag_to_delete = create_tag_for_asg_deletion(asg_name, 0)
+    autoscale = boto.connect_autoscale()
+    if len(autoscale.get_all_groups([asg_name])) < 1:
+        LOG.info("ASG {} no longer exists, will not remove deletion tag.".format(asg_name))
+    else:
+        autoscale.delete_tags([tag_to_delete])
```
Comment (macdiesel) [inline tubular/ec2.py:190]: Have you had a chance to try this against the ASG? While looking through the boto 2 docs it seems is the value of the tags don't match they will not be removed, though the documentation for the autoscale class isn't nearly as clear as the other docs.

Autoscale:
http://boto.cloudhackers.com/en/latest/ref/autoscale.html#boto.ec2.autoscale.AutoScaleConnection.delete_tags

Other delete tags:
http://boto.cloudhackers.com/en/latest/ref/ec2.html#boto.ec2.connection.EC2Connection.delete_tags

```
tags (dict or list) – Either a dictionary containing name/value pairs or a list containing just tag names. If you pass in a dictionary, the values must match the actual tag values or the tag will not be deleted. If you pass in a value of None for the tag value, all tags with that name will be deleted.
```

TLDR: We should probably set the tag value to None.

**Review by macdiesel — COMMENTED** [review]
@doctoryes Pass complete.

File: scripts/rollback_asg.py (hunk 4 above)
Comment (doctoryes) [reply scripts/rollback_asg.py]: Will do. This script has other problems which I'll fix as well.

File: tubular/asgard.py (hunk 6 above)
Comment (doctoryes) [reply tubular/asgard.py]: Will do.

File: tubular/asgard.py (hunk 7 above)
Comment (doctoryes) [reply tubular/asgard.py]: Will do.

File: tubular/asgard.py (hunk 5 above)
Comment (doctoryes) [reply tubular/asgard.py]: I decided yes. The code is trying to disable as many ASGs as possible in the list - it shouldn't stop attempting to disable other ASGs if one can't be disabled.

File: tubular/asgard.py [hunk 10]
```diff
@@ -490,78 +545,150 @@ def deploy(ami_id):
...
+                # Disable the ASG which failed first.
+                disable_cluster_asg(cluster, asg)
+                # Then disable any new other ASGs that have been newly enabled.
+                for cluster, asgs_to_disable in newly_enabled_asgs.iteritems():
+                    for asg in asgs_to_disable:
+                        try:
+                            disable_cluster_asg(cluster, asg)
+                        except:
+                            LOG.warning("Unable to disable ASG '{}' after failure.".format(asg))
+                            continue
```
Comment (doctoryes) [inline tubular/asgard.py]: This `continue` does nothing - I will remove it.

File: tubular/asgard.py (hunk 5 above)
Comment (doctoryes) [reply tubular/asgard.py]: My previous comment was wrong. A `continue` here would do nothing, just as the `continue` above does. So I won't add one here, since it won't change the program flow.

Comment (macdiesel) [reply tubular/asgard.py]: As long as it's consistent and moved in to a function it really doesn't matter. Let me ask though, do you think having the continue there makes the code more readable/understandable?

File: tubular/asgard.py (hunk 8 above)
Comment (maxrothman) [reply tubular/asgard.py:682]: I would prefer to decouple this step entirely and make it possible to have cleanup be in a separate pipeline stage to support use cases such as waiting for manual input before switching over, "soaking" new instances with no requests, etc.

Comment (macdiesel) [reply tubular/asgard.py:682]: Max, I don't think that manual input is something that would provide a lot of value here and goes against the grain of automate everything. 

That said, parsing your response it seems that perhaps soaking the instances may be a good idea?

Comment (maxrothman) [reply tubular/asgard.py:682]: Not necessarily? I'm having trouble articulating a good usecase for separating these two stages, but I can definitely imagine someone wanting to put additional pipeline stages in between them.

Comment (doctoryes) [reply tubular/asgard.py:682]: @maxrothman For this iteration of deploy and rollback, the goal is to automate our current release process - what DevOps currently does. While I could see useful reasons for keeping both ASGs active, such as directing a fractional amount of traffic to a new release indefinitely, it's my understanding that we're not currently doing that sort of thing. Nothing prevents us from implementing this in the future. But we'd like to just implement the same process that DevOps uses for now. How long does DevOps wait in the red/red state before disabling the old release's ASGs?

File: tubular/asgard.py (hunk 5 above)
Comment (doctoryes) [reply tubular/asgard.py]: With the code broken into a separate method, it's pretty clear without a `continue`.

File: tubular/ec2.py (hunk 9 above)
Comment (doctoryes) [reply tubular/ec2.py:190]: I have not tried tag removal from an actual ASG yet. But I will set the tag value to `None` here.

File: tubular/asgard.py (hunk 8 above)
Comment (doctoryes) [reply tubular/asgard.py:682]: I added a env-var-controlled delay between starting the new ASGs _and_ performing health checks on them.

**Comment by doctoryes** [comment]
@macdiesel Ready for another review.

File: tubular/asgard.py (hunk 8 above)
Comment (maxrothman) [reply tubular/asgard.py:682]: 👍

**Comment by macdiesel** [comment]
🚢 🇮🇹 👍
